src/inmanta/deploy/timers.py

```diff
@@ -0,0 +1,219 @@
+
+    def reset(self) -> None:
+        self.stop()
+
+        self.global_periodic_repair_task = None
+        self.global_periodic_deploy_task = None
+        self.periodic_repair_interval = None
+        self.periodic_deploy_interval = None
+
+    def stop(self) -> None:
+        for timer in self.resource_timers.values():
+            timer.cancel_next_schedule()
+
+        for task in [self.global_periodic_repair_task, self.global_periodic_deploy_task]:
+            if task:
+                self._sched.remove(task)
+
+    def join(self) -> None:
+        for timer in self.resource_timers.values():
+            timer.cancel_repair_task()
+
+    def initialize(self) -> None:
+        """
+        Set the periodic timers for repairs and deploys. Either per-resource if the
+        associated config option is passed as a positive integer, or globally if it
+        is passed as a cron expression string.
+        """
+        deploy_timer: int | str = agent_config.agent_deploy_interval.get()
+        repair_timer: int | str = agent_config.agent_repair_interval.get()
+
+        if isinstance(deploy_timer, str):
+            self.global_periodic_deploy_task = self.trigger_global_deploy(deploy_timer)
+        else:
+            self.periodic_deploy_interval = deploy_timer if deploy_timer > 0 else None
+
+        if isinstance(repair_timer, str):
+            self.global_periodic_repair_task = self.trigger_global_repair(repair_timer)
+        else:
+            self.periodic_repair_interval = repair_timer if repair_timer > 0 else None
+
+    def trigger_global_deploy(self, cron_expression: str) -> ScheduledTask:
+        """
+        Trigger a global deploy following a cron expression schedule.
+
+        :returns: the associated scheduled task.
+        """
+
+        cron_schedule = CronSchedule(cron=cron_expression)
+
+        async def _action() -> None:
+            await self._resource_scheduler.deploy(
+                reason=f"Global deploy triggered because of set cron {cron_expression}",
+                priority=TaskPriority.INTERVAL_DEPLOY,
+            )
+
+        task = self._sched.add_action(_action, cron_schedule)
+        assert isinstance(task, ScheduledTask)
+        return task
+
+    def trigger_global_repair(self, cron_expression: str) -> ScheduledTask:
+        """
+        Trigger a global repair following a cron expression schedule.
+
+        :returns: the associated scheduled task.
+        """
+        cron_schedule = CronSchedule(cron=cron_expression)
+
+        async def _action() -> None:
+            await self._resource_scheduler.repair(
+                reason=f"Global repair triggered because of set cron {cron_expression}",
+                priority=TaskPriority.INTERVAL_REPAIR,
+            )
+
+        task = self._sched.add_action(_action, cron_schedule)
+        assert isinstance(task, ScheduledTask)
+        return task
+
+    def update_resource(self, resource: ResourceIdStr, dirty: bool = False) -> None:
```

Comment:
I would move this closer to the `ensure_timer`

Reply:
What do you mean ? They are in 2 distinct classes at the moment. ( `ResourceTimer.ensure_timer` and `TimerManager.update_resource` )